### PR TITLE
Store something on each request. Update PoW types. Additional checks

### PR DESCRIPTION
**Pull request by forgetso**
We need to be able to see how many requests for Proof of Work captcha have been made

- [x] This PR changes the PoW generation mechanism to generate a record at the point of challenge creation. Various attributes are modified by subsequent requests to mark the stage the challenge is at (requested, userChecked, serverChecked).
- [x] The PoW types also receive a fairly big refactor.
- [x] The signatures are changed to be requestHash signatures for image captcha. Signature for PoW captcha remain as challenge (provider) and timestamp (user).
- [x] Types addded for get image challenge URL so that tests don't build if the URL changes
- [x] PoW and Image captcha records now inherit from a common type containing common statuses and a result object
- [x] Changes TranslationKeys to zod type for compatibility with zod schemas
- [x] add lastUpdatedTimestamp to captcha records so that scheduled tasks don't send the same data multiple times
- [x] store audit details of scheduled captcha sending task

### CODE REVIEW

File: packages/provider/src/api/captcha.ts [hunk 1]
```diff
@@ -35,6 +35,8 @@ import express, { type Router } from "express";
 import { Tasks } from "../tasks/tasks.js";
 import { handleErrors } from "./errorHandler.js";
 
+const NO_IP_ADDRESS = "NO_IP_ADDRESS";
```
Comment (goastler) [inline packages/provider/src/api/captcha.ts]: probably want to chuck 'as const' on the end of this for better typing

**Review by goastler — DISMISSED** [review]

File: dev/scripts/src/setup/setup.ts [hunk 2]
```diff
@@ -140,13 +139,13 @@ export async function setup(force: boolean) {
 
     env.logger.info(`Registering provider... ${defaultProvider.address}`);
 
-    defaultProvider.pair = await getPairAsync(network, providerSecret);
+    defaultProvider.pair = await getPairAsync(providerSecret);
 
     // If no PROSOPO_SITE_KEY is present, we will register a test account like //Eve.
```
Comment (HughParry) [inline dev/scripts/src/setup/setup.ts]: Stale comments are pretty confusing here

File: packages/api/src/api/ProviderApi.ts [hunk 3]
```diff
@@ -31,73 +30,69 @@ import {
...
-          [ApiParams.user]: {
-            [ApiParams.requestHash]: userRequestHashSignature,
-          },
-          [ApiParams.provider]: {
-            [ApiParams.timestamp]: providerTimestampSignature,
-          },
+    const body: CaptchaSolutionBodyType = {
+      [ApiParams.user]: userAccount,
+      [ApiParams.dapp]: this.account,
+      captchas,
```
Comment (HughParry) [inline packages/api/src/api/ProviderApi.ts]: all of these should be api params: blah

File: packages/provider/src/api/captcha.ts [hunk 4]
```diff
@@ -49,47 +52,45 @@ export function prosopoRouter(env: ProviderEnvironment): Router {
...
+        await tasks.imgCaptchaManager.getRandomCaptchasAndRequestHash(
+          datasetId,
+          user,
+          req.ip || NO_IP_ADDRESS,
         );
-      }
-    },
-  );
+      const captchaResponse: CaptchaResponseBody = {
+        captchas: taskData.captchas.map((captcha: Captcha) => ({
```
Comment (HughParry) [inline packages/provider/src/api/captcha.ts]: [ApiParams.captchas] or something

File: packages/provider/src/tasks/powCaptcha/powTasks.ts [hunk 5]
```diff
@@ -75,57 +74,101 @@ export class PowCaptchaManager {
...
+    const challengeRecord =
+      await this.db.getPowCaptchaRecordByChallenge(challenge);
+
+    if (!challengeRecord) {
+      logger.debug("No record of this challenge");
+      // no record of this challenge
+      return false;
+    }
+
+    if (challengeRecord.ipAddress !== ipAddress) {
```
Comment (HughParry) [inline packages/provider/src/tasks/powCaptcha/powTasks.ts]: If we could drop this for now please

File: packages/provider/src/tasks/powCaptcha/powTasksUtils.ts [hunk 6]
```diff
@@ -31,19 +31,8 @@ export const checkPowSolution = (
...
-        ERROR: "Captcha solution is invalid",
-        failedFuncName: checkPowSolution.name,
-        nonce,
-        challenge,
-        difficulty,
-      },
-    });
-  }
+): boolean => {
+  return validateSolution(nonce, challenge, difficulty);
```
Comment (HughParry) [inline packages/provider/src/tasks/powCaptcha/powTasksUtils.ts]: can we drop this func now that it just passes to a different one?

File: packages/provider/src/tasks/powCaptcha/powTasksUtils.ts [hunk 7]
```diff
@@ -72,15 +61,6 @@ export const checkPowSignature = (
...
-    throw new ProsopoContractError("CAPTCHA.INVALID_TIMESTAMP", {
-      context: {
-        ERROR: `Timestamp in which the Provider was selected must be within the last ${timeout / 1000} seconds`,
-        failedFuncName: checkRecentPowSolution.name,
-        challenge,
-      },
-    });
-  }
+): boolean => {
+  return verifyRecency(challenge, timeout);
```
Comment (HughParry) [inline packages/provider/src/tasks/powCaptcha/powTasksUtils.ts]: can we drop this func now that it just passes to a different one?

**Review by HughParry — COMMENTED** [review]
few comments

File: dev/scripts/src/setup/setup.ts (hunk 2 above)
Comment (forgetso) [reply dev/scripts/src/setup/setup.ts]: Removed

File: packages/provider/src/tasks/powCaptcha/powTasks.ts (hunk 5 above)
Comment (forgetso) [reply packages/provider/src/tasks/powCaptcha/powTasks.ts]: Dropped

**Review by HughParry — APPROVED** [review]

**Pull request merged**